examples/pytorch/question-answering/sparseml_utils.py

```diff
@@ -0,0 +1,205 @@
+    """
+    Question Answering trainer using distilation with customized optimizer using SparseML
+
+    :param nm_prune_config: recipe for model sparsification
+    :param teacher: teacher model
+    :param distill_hardness: weight of the teacher loss
+    :param temperature: temperature used for loss
+    :param args, kwargs: arguments passed into parent class
+    """
+
+    def __init__(self, nm_prune_config, teacher=None, distill_hardness=0.5, temperature=2.0, *args, **kwargs):
+        super().__init__(nm_prune_config, *args, **kwargs)
+        self.teacher = teacher
+        self.distill_hardness = distill_hardness
+        self.temperature = temperature
+        self.criterion = torch.nn.CrossEntropyLoss()
+
+    def compute_loss(self, model, inputs, return_outputs=False):
+        """
+        Computing loss using teacher/student distillation
+        """
+        outputs = model(**inputs)
+        loss = outputs["loss"]
+        if self.teacher is not None:
+            input_device = inputs["input_ids"].device
+            self.teacher = self.teacher.to(input_device)
+            start_logits_student = outputs["start_logits"]
+            end_logits_student = outputs["end_logits"]
+            start_logits_label = inputs["start_positions"]
+            end_logits_label = inputs["start_positions"]
+            with torch.no_grad():
+                teacher_output = self.teacher(
+                    input_ids=inputs["input_ids"],
+                    token_type_ids=inputs["token_type_ids"],
+                    attention_mask=inputs["attention_mask"],
+                )
+            start_logits_teacher = teacher_output["start_logits"]
+            end_logits_teacher = teacher_output["end_logits"]
+            loss_start = (
+                F.kl_div(
+                    input=F.log_softmax(start_logits_student / self.temperature, dim=-1),
+                    target=F.softmax(start_logits_teacher / self.temperature, dim=-1),
+                    reduction="batchmean",
+                )
+                * (self.temperature ** 2)
+            )
+            loss_end = (
+                F.kl_div(
+                    input=F.log_softmax(end_logits_student / self.temperature, dim=-1),
+                    target=F.softmax(end_logits_teacher / self.temperature, dim=-1),
+                    reduction="batchmean",
+                )
+                * (self.temperature ** 2)
+            )
+            teacher_loss = (loss_start + loss_end) / 2.0
+            loss_start = self.criterion(start_logits_student, start_logits_label)
+            loss_end = self.criterion(end_logits_student, end_logits_label)
+            label_loss = (loss_start + loss_end) / 2.0
+            loss = ((1 - self.distill_hardness) * label_loss) + (self.distill_hardness * teacher_loss)
+        return (loss, outputs) if return_outputs else loss
+
+
+def convert_example_to_features(example, tokenizer, max_seq_length, doc_stride, max_query_length=30):
```

Comment:
is this the simplest way we can do this?

Additionally, we should add some comments to this code and likely break it out into multiple functions